Src/ILGPU/Runtime/ExchangeBuffer.cs

```diff
@@ -381,4 +458,682 @@ protected override void Dispose(bool disposing)
+        {
+            get => CPUView[index2];
+
+            set => CPUView[index2] = value;
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Sets the contents of the current buffer to zero.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        public override void MemSetToZero(AcceleratorStream stream) =>
+            Buffer.MemSetToZero(stream);
+
+        /// <summary>
+        /// Copies the current contents into a new byte array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        public override byte[] GetAsRawArray(AcceleratorStream stream) =>
+            Buffer.GetAsRawArray(stream);
+
+        /// <summary>
+        /// Copies the current contents into a new byte array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <param name="byteOffset">The offset in bytes.</param>
+        /// <param name="byteExtent">The extent in bytes (number of elements).</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        protected internal override ArraySegment<byte> GetAsRawArray(
+            AcceleratorStream stream,
+            Index1 byteOffset,
+            Index1 byteExtent) =>
+            Buffer.GetAsRawArray(stream, byteOffset, byteExtent);
+
+        /// <summary>
+        /// Copies the current contents into a new array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        public unsafe T[] GetAsArray(AcceleratorStream stream)
+        {
+            CopyFromAccelerator(stream);
+            stream.Synchronize();
+
+            var data = new T[Length];
+            fixed (T* ptr = &data[0])
+            {
+                System.Buffer.MemoryCopy(
+                    cpuMemoryPointer,
+                    ptr,
+                    LengthInBytes,
+                    LengthInBytes);
+            }
+            return data;
+        }
+
+        /// <summary>
+        /// Gets this buffer as a 2D jagged array from the accelerator
+        /// </summary>
+        /// <param name="stream">The stream to use</param>
+        /// <returns>A jagged array containing all the elements in this exchnage buffer</returns>
+        public unsafe T[][] GetAs2DArray(AcceleratorStream stream)
+        {
+            CopyFromAccelerator(stream);
+            stream.Synchronize();
+
+            var data = new T[Length];
+            fixed (T* ptr = &data[0])
+            {
+                System.Buffer.MemoryCopy(
+                    cpuMemoryPointer,
+                    ptr,
+                    LengthInBytes,
+                    LengthInBytes);
+            }
+
+            var data2D = new T[Extent.X][];
```

Comment:
Same here: this essentially does exactly the same like a `MemoryBuffer2D` instance does